packages/hardhat-core/src/internal/artifacts.ts

```diff
@@ -173,7 +174,90 @@ export class Artifacts implements IArtifacts {
     );
 
     const buildInfoPath = path.join(buildInfoDir, `${buildInfoName}.json`);
-    await fsExtra.writeJson(buildInfoPath, buildInfo, { spaces: 2 });
+
+    // JSON.stringify of the entire build info can be really slow
+    // in larger projects, so we stringify per part and incrementally create
+    // the JSON in the file.
+    //
+    // We split this code into different curly-brace-enclosed scopes so that
+    // partial JSON strings get out of scope sooner and hence can be reclaimed
+    // by the GC if needed.
+    const file = await fsPromises.open(buildInfoPath, "w");
+    try {
+      {
+        const withoutOutput = JSON.stringify({
+          ...buildInfo,
+          output: undefined,
+        });
+
+        // We write the JSON (without output) except the last }
+        await file.write(withoutOutput.slice(0, -1));
+      }
+
+      {
+        const outputWithoutSourcesAndContracts = JSON.stringify({
+          ...buildInfo.output,
+          sources: undefined,
+          contracts: undefined,
+        });
+
+        // We start writing the output
+        await file.write(',"output":');
+
+        // Write the output object except for the last }
+        await file.write(outputWithoutSourcesAndContracts.slice(0, 1));
```

Comment:
Uhm, shouldn't this have caused some test to fail?

```suggestion
        await file.write(outputWithoutSourcesAndContracts.slice(0, -1));
```

Reply:
Maybe in the tests this object is always {} so this was working by coincidence?

Reply:
Added enough tests to uncover this and similar bugs (none was present) and fixed it.